flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -428,6 +431,123 @@ abstract class TableEnvironment(val config: TableConfig) {
+    * @param logicalRowType the logical type with correct field names (esp. for POJO field mapping)
+    * @param expectedTypeInfo the outptu type of the sink
+    * @param functionName name of the map function. Must not be unique but has to be a
+    *                     valid Java class identifier.
+    */
+  protected def sinkConversion[T](
+      physicalRowTypeInfo: TypeInformation[Row],
+      logicalRowType: RelDataType,
+      expectedTypeInfo: TypeInformation[T],
+      functionName: String)
+    : Option[MapFunction[Row, T]] = {
+
+    // validate that at least the field types of physical and logical type match
+    // we do that here to make sure that plan translation was correct
+    val logicalRowTypeInfo = FlinkTypeFactory.toInternalRowTypeInfo(logicalRowType)
+    if (physicalRowTypeInfo != logicalRowTypeInfo) {
+      throw TableException("The field types of physical and logical row types do not match." +
+        "This is a bug and should not happen. Please file an issue.")
+    }
+
+    // expected type is a row, no conversion needed
+    // TODO this logic will change with FLINK-5429
+    if (expectedTypeInfo.getTypeClass == classOf[Row]) {
+      return None
+    }
+
+    // convert to type information
+    val logicalFieldTypes = logicalRowType.getFieldList.asScala map { relDataType =>
+      FlinkTypeFactory.toTypeInfo(relDataType.getType)
+    }
+    // field names
+    val logicalFieldNames = logicalRowType.getFieldNames.asScala
+
+    // validate expected type
+    if (expectedTypeInfo.getArity != logicalFieldTypes.length) {
+      throw new TableException("Arity of result does not match expected type.")
+    }
+    expectedTypeInfo match {
+
+      // POJO type expected
+      case pt: PojoTypeInfo[_] =>
+        logicalFieldNames.zip(logicalFieldTypes) foreach {
+          case (fName, fType) =>
+            val pojoIdx = pt.getFieldIndex(fName)
+            if (pojoIdx < 0) {
+              throw new TableException(s"POJO does not define field name: $fName")
+            }
+            val expectedTypeInfo = pt.getTypeAt(pojoIdx)
+            if (fType != expectedTypeInfo) {
+              throw new TableException(s"Result field does not match expected type. " +
+                s"Expected: $expectedTypeInfo; Actual: $fType")
+            }
+        }
+
+      // Tuple/Case class type expected
+      case ct: TupleTypeInfoBase[_] =>
+        logicalFieldTypes.zipWithIndex foreach {
+          case (fieldTypeInfo, i) =>
+            val expectedTypeInfo = ct.getTypeAt(i)
+            if (fieldTypeInfo != expectedTypeInfo) {
+              throw new TableException(s"Result field does not match expected type. " +
+                s"Expected: $expectedTypeInfo; Actual: $fieldTypeInfo")
+            }
+        }
+
+      // Atomic type expected
+      case at: AtomicType[_] =>
+        if (logicalFieldTypes.size != 1) {
+          throw new TableException(s"Result does not have a single field, " +
```

Comment:
Improve error message: "Requested result type is an atomic type but result has more than a single field."